File: app/src/main/java/com/maxwellscore/sparkone/domain/WeatherDataMapper.java
package com.maxwellscore.sparkone.domain;


import static com.maxwellscore.sparkone.data.DataConstants.CLOUDY_TYPE;
import static com.maxwellscore.sparkone.data.DataConstants.RAIN_TYPE;
import static com.maxwellscore.sparkone.data.DataConstants.SUNNY_TYPE;

import com.maxwellscore.sparkone.data.dto.TemperatureData;
import com.maxwellscore.sparkone.data.dto.WeatherData;
import com.maxwellscore.sparkone.domain.entities.Weather;
import com.maxwellscore.sparkone.domain.entities.WeatherType;


public class WeatherDataMapper {

    public Weather dataToDomain(WeatherData data) {
        TemperatureData temperatureData = data.getTemperatureData();
        String temperature = String.format(
                "%s°%s",
                temperatureData.getTemperature(),
                temperatureData.getUnit()
        );
        return new Weather(data.getCityName(), temperature, getWeatherTypeFromString(data.getWeatherType()));
    }

    private WeatherType getWeatherTypeFromString(String type) {
        WeatherType result;
        switch (type) {
            case CLOUDY_TYPE:
                result = WeatherType.CLOUDY;
                break;
            case SUNNY_TYPE:
                result = WeatherType.SUNNY;
                break;
            case RAIN_TYPE:
                result = WeatherType.RAIN;
                break;
            default:
                throw new IllegalArgumentException("Такого типа погоды не существует");
        }
        return result;
    }
}

File: kotlinexample/src/main/java/com/maxwellscore/kotlinexample/domain/WeatherDataMapper.kt
package com.maxwellscore.kotlinexample.domain

import com.maxwellscore.kotlinexample.data.DataConstants
import com.maxwellscore.kotlinexample.data.dto.WeatherData
import com.maxwellscore.kotlinexample.domain.entities.Weather
import com.maxwellscore.kotlinexample.domain.entities.WeatherType

class WeatherDataMapper {
    fun dataToDomain(data: WeatherData): Weather {
        val temperatureData = data.temperatureData
        val temperature = "${temperatureData.temperature}°${temperatureData.unit}"
        return Weather(data.cityName, temperature, getWeatherTypeFromString(data.weatherType))
    }

    private fun getWeatherTypeFromString(type: String) = when (type) {
        DataConstants.CLOUDY_TYPE -> WeatherType.CLOUDY
        DataConstants.SUNNY_TYPE -> WeatherType.SUNNY
        DataConstants.RAIN_TYPE -> WeatherType.RAIN
        else -> throw IllegalArgumentException("Такого типа погоды не существует")
    }
}